### ISSUE TITLE
Postgresql Replication Proxy Implementation

### ISSUE THREAD

**Issue by mstrofbass** [freelawproject/courtlistener#949]
Figured it's probably better to separate this out into a new issue than piggy back on the more general "what's the best solution" issue.

There are four main components involved in this implementation:

1. Publisher Database
2. Elastic Load Balancer (Network)
3. EC2 Proxy Instance
4. Subscriber Database

There are several additional components/things needed as well:

1. SSL Certificate 
2. Hostname/Elastic IP for the ELB

Essentially, the publisher database is configured to replicate to the subscriber database by pointing the publisher config to the hostname used for the ELB. This hostname is an A record, thus eliminating the reverse DNS issues experienced by trying to use the CNAME record used by Amazon RDS.

The ELB/EC2 instance configuration is about as simple as it gets: the ELB accepts traffic on port 5433 and then forwards the traffic to port 5433 of the EC2 proxy instance, which then forwards the traffic to port 5433 of the subscriber database. We could eliminate the ELB and go straight to the EC2 instance, but the ELB makes dealing with SSL certs incredibly simple (and allows us to easily add EC2 instances for failover purposes if we want).

## Config Details

I'll include some high level config details below and then describe the low level details for each individual component in separate replies.

### Publisher Database

The configuration for the publisher database probably won't have to change beyond updating the target address to the URL/IP used for the ELB. I have not actually set up PG replication so there may be more to it than this.

### ELB

This is an AWS Network Elastic Load Balancer. The "Network" ELB works at the TCP level (transport layer?) while the "Application" ELB works at the HTTP layer (application layer?). The ELB is set to only listen on TCP port 5433 and forward traffic to port 5433 of any instance in the target group assigned to the ELB. 

For AWS ELBs, the target group basically just defines the set of instances (or IPs) to which the ELB can/will forward traffic. The current plan is to just have a single instance assigned to the target group. If another instance is desired (e.g., for failover purposes), it merely needs to be assigned to the target group. 

### EC2 Instance

It is not clear what the appropriate sizing is, but we're going to start with a t2.micro. Given the high performance of the proxy, which is literally just forwarding TCP packets, it will not be surprising if the t2.micro is plenty sufficient.

I initially started with Amazon's Linux distro, but finding the repo for the most recent LTS stable version of HAProxy was more complicated than needed, so I decided to just go with Ubuntu. For this implementation, any distro should be fine.

#### HAProxy

The HAProxy configuration is about as simple as it gets. The config is the basic default config (which is probably more complex than necessary) with a listener configured to listen on port 5433 and send to the AWS RDS endpoint URL corresponding to the subscriber. 

### Subscriber Database

The replication configuration for the subscriber database may need to be updated but I'm not entirely sure how.

**Comment by mlissner** [comment]
> Figured it's probably better to separate this out into a new issue than piggy back on the more general "what's the best solution" issue.

Either way, it's all good. But probably worth at least mentioning issue #938 so we're cross linked. :)

-------

A couple thoughts:

1. Postgres calls these the publisher and subscriber, so I updated your terminology above. In general the config of these works like this:

     - You [CREATE][1] a publication on the publisher.
     - You [CREATE][2] a subscription on the subscriber.

    The subscription has information about where the publisher is, including the hostname or IP address, so once this is all set up, we'll want to run an [ALTER SUBCRIPTION][3] command to tweak the currently set IP address so it's a hostname.

1. We're currently using AWS's network firewall (I forget the product name) to only allow traffic to the publisher from IP addresses that we trust. One step of this will presumably be to allow traffic to the ELB from *only* those IP addresses.

2. Re HAProxy, I think my concerns are:

    - Logging — We don't want logs piling up forever. Can we make sure we've got log rotation enabled?
    - Network security — can  we make sure that only the ELB and the RDS instance can access this guy?
    - Security — Can we make sure that security updates are automatically applied?
    - Anything else here that we'd want to do to make sure this server never needs us ever again?

This sounds really solid overall. Thank you! 

[1]: https://www.postgresql.org/docs/10/sql-createpublication.html
[2]: https://www.postgresql.org/docs/11/sql-createsubscription.html
[3]: https://www.postgresql.org/docs/10/sql-altersubscription.html

**Comment by mstrofbass** [comment]
1. Logging: HAProxy logrotation is enabled by default in Ubuntu. I don't really know anything about how this works, so I'm not sure if that's an artifact of the apt installation or the OS doing it automatically when something writes logs to /dev/log. The default config for the HAProxy log rotation is as follows:

`/var/log/haproxy.log {
    daily
    rotate 52
    missingok
    notifempty
    compress
    delaycompress
    postrotate
        invoke-rc.d rsyslog rotate >/dev/null 2>&1 || true
    endscript
}
`

My understanding is that `rotate 52` means it'll keep 52 time periods worth of files (in this case daily). I think daily is probably the correct period but I'm not sure how long you want to keep them around. I also don't know how large they will get, which contributes to that uncertainty. Perhaps these are okay defaults and we check it after a couple of weeks?

2. Network Security: So my dry run began with a security group applied to the EC2 instance that only allowed traffic to hit the EC2 instance if it came from an IP on the same subnet. It wasn't until I started allowing my IP that I was able to start hitting the EC2 instance through the ELB. I realized that the traffic still appears to be coming from the client, not the ELB, which seems obvious in hindsight. So the security group rules, which are applied per instance, not on the ELB itself, would block my traffic even if I had traffic from the ELB open.

So what this means is that we can apply security group rules to the EC2 instance to only allow traffic from the sources we want. This is the same effect as what you want, but technically is not the same. There is a possibility that we can put a firewall in front of the ELB, but I'm not familiar with that configuration.

The only firewall product from Amazon I'm familiar with is the Web Application Firewall (WAF). Is that the one you're using? A previous project I worked on used that in front of an Application Load Balancer.

3. Security - Ubuntu has an unattended-upgrades package that handles security updates. I'm not familiar with it, so I'll do some digging. My main question pertains to updates requiring restarts.

4. The main thing that comes to mind is monitoring alerts...which is less about making sure it will never need us again and more about making sure we know if it ever needs us. For this server, the two things I'd be worried about is CPU usage and disk space. The former is easy (although I have no idea what a sane default would be for our scenario...might just have to wing it and tweak it as we go). The latter I would guess is easy but I'm not familiar with the best way to handle it. It has to be handled on the server (either directly or by using the extended monitoring functionality AWS offers) and would primarily cover a scenario in which traffic jumped to a degree that the log rotation settings were insufficient given the disk size. 

I'll follow up on everything that needs to be followed up re: this stuff a bit later.

**Comment by mstrofbass** [comment]
# HAProxy Config

This is really the only config file that's necessary for the core infrastructure:

```
global
        log /dev/log    local0
        log /dev/log    local1 notice
        chroot /var/lib/haproxy
        stats socket /run/haproxy/admin.sock mode 660 level admin expose-fd listeners
        user haproxy
        group haproxy
        daemon

defaults
        log     global
        mode    tcp
        option  tcplog
        timeout connect 5000
        timeout client  50000
	timeout client-fin 30s
        timeout server  50000
        timeout tunnel  1h
        errorfile 400 /etc/haproxy/errors/400.http
        errorfile 403 /etc/haproxy/errors/403.http
        errorfile 408 /etc/haproxy/errors/408.http
        errorfile 500 /etc/haproxy/errors/500.http
        errorfile 502 /etc/haproxy/errors/502.http
        errorfile 503 /etc/haproxy/errors/503.http
        errorfile 504 /etc/haproxy/errors/504.http

listen pg-repl-in
	bind *:5432
        bind *:5433
        server pg_rds *RDS hostname* check port 5432
```

This is a very basic configuration. In the global section we set the log info, where HAProxy will chroot itself, some stats stuff I don't know how to use, user/group, and setting it to run as a daemon.

Defaults are fairly self-explanatory. A few details:

1. timeout tunnel - applies to connection in which a tunnel is created. I'm not sure if this applies to the postgres replication process.
2. timeout client-fin - this has something to do with a scenario in which a connection is half-closed which can occur when one side of a tunnel closes the connection and the other side has not. This is a bit beyond my TCP knowledge but is an issue that can occur during longer running tunnel connections.

The "listen" section just defines the actual proxy. In this case we bind to the two ports and any request to those two ports gets forwarded to the server, which is defined by the RDS hostname. We also tell it to use port 5432 as the port to check health status.

**Comment by mstrofbass** [comment]
The best way for me to detail the AWS config is just to do a step-by-step instructions for the configuration based on my dry run. Note that it can't be assumed this will work perfectly as it was documented while troubleshooting failed configs and then updated to apply to a setup using SSL/TLS (which I wasn't able to directly test). I also just tested by using a client to connect to the RDS instance instead of setting up a fully replicated database.

It's also worth noting that none of this is "destructive" inasmuch as this whole thing can be built out without impacting the existing infrastructure (until the switch is flipped for the replication, of course). 

1. General Notes:

    - All infrastructure should be in the same VPC and (probably) the same availability zone (AZ) (at least until failover functionality is implemented)
    - b. I specify names in the config below but those can be changed to whatever is desired, obviously.

2. Set up SSL certificate via AWS Certificate Manager - I am not sure what the verification requirements are for this, so I don’t know how long it will take and am not sure what the specific steps are (I didn’t do a dry run for this)

3. Security Group – You should be able to use the security group you currently use for direct access to the RDS or, if planning on using a firewall, this is unnecessary.

4. Create haproxy EC2 instance
    - VPC: VPC containing the RDS instance
    - AZ: Desired AZ, which will include all the additional infrastructure
    - Security Group: As mentioned above or none if using firewall

5. Create Target Group for port 5432
    - Name: repl-proxy-5432
    - Protocol: TLS
    - Port: 5432
    - Target Type: Instance
    - VPC: VPC containing the RDS
    - Add haproxy EC2 instance to target group

6. Create Target Group for port 5433
    - Name: repl-proxy-5432
    - Protocol: TLS
    - Port: 5433
    - Target Type: Instance
    - VPC: VPC containing the RDS
    - Add haproxy EC2 instance to target group

7.	Create Network ELB
    - Internet Facing
    - Listeners:
        - TLS; port 5432
        - TLS; port 5433
        - Optional: Same ports for TCP if SSL-less support is desired. This may be unnecessary though; specifying “TLS” as the protocol may not force it to use TLS (or, to put it another way, specifying TLS may still let it fall back to standard TCP).
    - VPC and availability zone as above
    - Security Settings:
        - Select certificate from above
        - Default security policy (ELBSecurityPolicy-2016-08) should be fine unless there are specific requirements for which TLS versions must be used.
        - See for more details: https://docs.aws.amazon.com/elasticloadbalancing/latest/application/create-https-listener.html#describe-ssl-policies
    - Target Group
        - Repl-proxy-5432 for port 5432
        - Repl-proxy-5433 for port 5433

**Comment by mstrofbass** [comment]
Clearly I did not include the instructions for setting up HAProxy...that was the first thing I did but was fairly easy and I can document a bit later.

**Comment by mlissner** [comment]
I'm loving your detailed journaling of this. Thank you very much.

> Logging

Our privacy policy is 12 weeks, but this doesn't have user info, so the default is fine. Just wanted to make sure *something* was in place after HAProxy was installed. Thanks for verifying.

> Network security.

So if I understand correctly, the ELB forwards the traffic to the EC2 instance with the IP of the client, not the IP of the ELB. That's...weird, but fine, I guess. And so if we want to block traffic to the ELB, we have to put something in front of *it*, which means probably using Web Application Firewall (WAF)?

We're using security groups to handle the network traffic blocking, and can use that to block access to the EC2 instance. I don't know how that'd be different than blocking access to the ELB using WAF (a product I've never looked at). So yeah...unless you see a reason otherwise, I say we keep WAF out of the picture, use security groups for EC2 (as we do for RDS currently), and I think we should be good?

> My main question pertains to updates requiring restarts.

Enable 'em. That'll cause an occasional momentary blip for folks, but I think that's fine and I'd rather a small blip than an insecure box that we've forgotten existed.

> Monitoring alerts

Good point. Let's make a todo:

 - [ ] Set up disk and CPU alerts for the EC2 instance using AWS monitoring tools

I agree that we'll want to just monitor this as we go.

> HAProxy config

I've never configured this, but it's fun to see that they invented their own config style. It's like a rite of passage or something. Honestly folks, use JSON or YAML, if you must. 

Anyway, it all looks good to me. Is it worth commenting out the stats part if we'll never use that? Just put a note that says it's available if we want it? Looks like something that could open a hole into the server that we don't know or understand and may as well keep closed.

> The best way for me to detail the AWS config is just to do a step-by-step instructions for the configuration based on my dry run.

I wondered how we'd work that. This is great though. Documentation, baby.

-----------

What's left to do? This looks nearly ready to launch?

**Comment by mstrofbass** [comment]
> So if I understand correctly, the ELB forwards the traffic to the EC2 instance with the IP of the client, not the IP of the ELB. That's...weird, but fine, I guess. 

At the TCP-level, yes. One of the lower level protocols will likely identify the ELB (or whatever device was last; I think this would be the Ethernet frame or equivalent, but I dig into that kind of stuff so sporadically that I never can remember my network stack).

If the ELB modified the TCP packet (or whatever the term is), then nothing after the TCP packet would know where the packet origin is, which would prevent us from, say, using a firewall on the server itself. It would also prevent HAProxy from routing based on the IP (e.g., for stickiness...although I'm guessing that will use cookies most of the time).

> And so if we want to block traffic to the ELB, we have to put something in front of it, which means probably using Web Application Firewall (WAF)?

> We're using security groups to handle the network traffic blocking, and can use that to block access to the EC2 instance. I don't know how that'd be different than blocking access to the ELB using WAF (a product I've never looked at). So yeah...unless you see a reason otherwise, I say we keep WAF out of the picture, use security groups for EC2 (as we do for RDS currently), and I think we should be good?

I concur that using security groups is probably the best route. 

> I've never configured this, but it's fun to see that they invented their own config style. It's like a rite of passage or something. Honestly folks, use JSON or YAML, if you must.

Don't forget that many of these products were initially developed in like the 90s when there wasn't much in the way of standardization (you see this with most of the older applications like Apache and NGINX). And no one wants to rewrite the config parsing portions of the code. :P

> Anyway, it all looks good to me. Is it worth commenting out the stats part if we'll never use that? Just put a note that says it's available if we want it? Looks like something that could open a hole into the server that we don't know or understand and may as well keep closed.

I don't think there's a concern either way, so let's just get rid of it for now.

---

> What's left to do? This looks nearly ready to launch?

Figuring out how we want to test and then switch over to this new infrastructure, which will be best handled by you since I (1) don't know much about PGSQL replication and (2) don't have a complete understanding of how the existing infrastructure is set up and thus don't know what potential impact this might have.

As mentioned above, building the infrastructure has no impact on the existing infrastructure and can co-exist in parallel for as long as we want. So the general implementation plan is:

1. Build out the infrastructure.
2. Test access using a postgresql client.
3+. ???

Once we verify that it's proxying correctly for the client, we need to test it with the replication. This can, obviously, be done either using a test database or the live database. 

Testing with a test database is obviously the best choice, but I'm not sure what the level of effort is for setting up a test database that will actually be sufficient to test replication. I only mention testing with the live database because I assume that the worst thing that can happen is that replication fails and most replication is designed to gracefully handle failures, so the potential impact is limited.

Perhaps the best option is this:

1. Set up a new subscriber RDS instance with the same specs as the current subscriber RDS instance.
2. Configure the new subscriber RDS instance to replicate from the same publisher as the current subscriber RDS instance.
3. Validate that it is replicating as intended (we can validate against the current subscriber RDS instance).
4. Migrate all external users to the new subscriber RDS instance.

I'm not sure how many people are subscribed to the RDS instance now, but this plan will allow us to monitor everything as people are switched over to the new RDS instance and, if necessary, switch back to the old instance if problems are encountered.

**Comment by mstrofbass** [comment]
To clarify, we can start setting up the infrastructure as soon as you want (time permitting). The only thing up in the air is the timeframe for actually switching to the new infrastructure.

**Comment by mlissner** [comment]
Sorry to keep you waiting here. This plan sounds great. I'll hit you up on Slack to make a plan to execute this.

**Comment by mlissner** [comment]
I believe this is now in place with one of our clients (I'm monitoring now). Now it's onwards to updating everything else. Right now the architecture is:

    CL master server replicates to---->
        AWS replica, which replicates to----->
            Client servers

The middle hop now has a proper DNS name, so the next step is to update both it's inbound and outbound replication connections.

**Comment by mlissner** [comment]
I believe the way to fix this is to copy the results from this:

```sql
select * from pg_subscription;
```

Then use them to make and run this:

```sql
ALTER SUBSCRIPTION xxx DISABLE;
ALTER SUBSCRIPTION xxx CONNECTION 'host=replication.courtlistener.com port=5432 password=xxx user=xxx dbname=courtlistener';
ALTER SUBSCRIPTION xxx ENABLE;
```

**Comment by mlissner** [comment]
That seems to have worked well. Onwards to the next client, then we close this thing down.

**Comment by mlissner** [comment]
Next client is done too. That's it. I regret how easy this was and how long I put it off. It's been on my short to do list now since May.

**Comment by mlissner** [comment]
Note issue #1123 for what needs to happen when the IP address changes.
